### PR TITLE
Better release notes

### PR DESCRIPTION
**Pull request by AronNovak**
#172 

## Status
image

### CODE REVIEW

**Comment by AronNovak** [comment]
Sample:

## Changelog
- Automatic deployment - never go back in time [3h] (#138)
  - Deployment fix (#156)
- Fix failing test on master (#142)
  - Fix travis on non-autodeploy repo (#143)
- Check DrupalStan [3h] (#107)
  - PHPStan + DrupalStan (#118)
- Migrate Users and Article [5h] (#108)
  - Migrate default users and content (#112)
- Keep Pantheon's setting.php in the repo [2h] (#80)
  - Pantheon settings in GH repo (#106)
- Add Redis (#79)
  - Redis (#81)
- Pure Robo release notes generator (#70)
  - Release notes documentation (#73)
  - Release notes based on Robo (#71)
- Add oomphinc/composer-installers-extender (#64)
  - Add oomphinc/composer-installers-extender (#65)
- Port recent improvements from client project [1h] (#50)
  - More helpful auto-depoy-config output (#52)
  - Port RoboFile improvements from client project (#51)
- Travis indicates success even if ddev failed (#36)
  - Better error handling in Robo (#38)
- Configure composer.json to download libraries (#32)
  - Prepare composer.json to download libraries from npm (#33)
- Rename theme name from `theme_server` (#17)
  - Rename theme (#21)
- More robust rollback protection (#168)
- Bump lodash from 4.17.20 to 4.17.21 in /web/themes/custom/server_theme (#158)
- Bump postcss from 8.2.2 to 8.2.10 in /web/themes/custom/server_theme (#159)
- Bump glob-parent from 5.1.1 to 5.1.2 in /web/themes/custom/server_theme (#165)
- Drupal 9 upgrade (#149)
- CI code cleanup (#161)
- Upgrade Elasticsearch (#152)
- Different containers in styleguide (#121)
- Add generator for photographic placeholder image (#124)
- Travis optimization (#141)
- Ubuntu focal - to avoid the need to upgrade Docker (#139)
- Change deploy:pantheon-install-env command to not double install on QA (#137)
- Elasticsearch provisioning - user rework (#131)
- Gcr Docker mirror to avoid download failures (#136)
- Is master failing? (#134)
- Elasticsearch updates (#129)
- Use picsum service for image placeholders (#127)
- Drupal core upgrade (#126)
- Pulling DB & Files From Pantheon (#120)
- Whitelist is now called Safelist, update config (#125)
- Update Pluggable entity view builder (PEVB) module to 1.1.1 (#123)
- Search test fix (#119)
- Migrate redesign (#1)
- Update README.md (#115)
- Rename scss to pcss for better IDE integration (#114)
- Render placeholder image with text (#113)
- Elasticsearch - fix entity deletion / unpubllish (#111)
- Fix CSS compile (#110)
- Simplify and cleanup CSS purge (#109)
- Set memory limit to be like the one in Pantheon (#105)
- Upgrade ddev to 1.16.4 (#104)
- Generic theme (#103)
- Regenerate config with ddev 1.16.2 (#101)
- Composer 2 (#97)
- Release notes beauty fix (#95)
- Robo command to deploy a release tag (#94)
- Remove the cache clears before a database update (#96)
- Use 1.0.0 version of Pluggable entity view builder (#91)
- Security update - Drupal core (#93)
- Tailwind 2 with PostCSS (#88)
- Elasticsearch_connector - date facet fix (#92)
- Update Drupal 8 core and elastic-connector (#90)
- Add missing server_theme_content__image_and_teaser (#89)
- Make imagemin tools available inside DDEV (#85)
- Autodeploy robustness (#84)
- Keep footer sticky to the bottom (#78)
- Update hero header dimensions css (#76)
- Compress the SVGs while compiling the theme. (#69)
- Install command follow-up (#72)
- Add a robo command to reinstall an environment on Pantheon. (#68)
- Elasticsearch (#67)
- Freeze coder module inside Travis (#66)
- Use Tailwind's purge whitelist instead of the purgecss.html.twig workaround file (#62)
- Update running workflows filter when deploying to pantheon (#60)
- Gitkeep for empty image dir to make sure theme can be compiled (#58)
- Fix function signature (#56)
- Always display PHP errors / notices (#57)
- Bump lodash from 4.17.15 to 4.17.19 in /web/themes/custom/server_theme (#53)
- Add Entity View Build plugins (#55)
- Upgrade core to 8.9.3 (#54)
- Tailwind integration, Style guide and NodeViewBuilders (#49)
- Upgrade ddev to 1.15.1 (#48)
- Basic perf settings (#44)
- Use local PHPCS (#43)
- Prevent rewriting development.services.yml (#42)
- Ignore node modules on phpcs (#41)
- More verbose deployment failure (#39)
- Mkcert fix (#35)
- Debug DDEV startup / action failure (#34)
- Add ddev exec command wrapper to terminus (#31)
- Add link to Pantheon section in README (#30)
- Add ddev command wrapper to terminus (#29)
- Automatic deployment integration (#24)
- Update Readme instructions for Pantheon. (#20)
- Move gitignore from theme to base. (#19)


## Contributors
- @AronNovak
- @amitaibu
- @mariano-dagostino
- @bboro 


## Code statistics
- Lines added: 23052
- Lines deleted: 12463
- Files changed: 399

File: README.md [hunk 1]
```diff
@@ -153,7 +153,14 @@ To Deploy to a Pantheon environment (e.g. TEST or LIVE) you can use
 ### Release notes
 
 Deployments should imply a release, you can generate a release notes based on
-tags. You can generate a changelog using
+tags.
+In order to provide verbose release notes, it is required to [create a personal
```
Comment (amitaibu) [inline README.md:157]: :+1:

File: RoboFile.php [hunk 2]
```diff
@@ -926,11 +930,38 @@ public function generateReleaseNotes($tag = NULL) {
...
+    if (!empty($remote)) {
+      $origin_parts = preg_split('/[:\/]/', str_replace('.git', '', $remote));
+      if (!empty($origin_parts[1]) && !empty($origin_parts[2])) {
+        $github_org = $origin_parts[1];
+        $github_project = $origin_parts[2];
+      }
+    }
+
+    if (!isset($github_org) || !isset($github_project)) {
+      $this->say('No origin remote is configured, not trying to fetch details from GitHub API then');
```
Comment (amitaibu) [inline RoboFile.php:948]:
```suggestion
      $this->say('No GitHub project or GitHub organization found, so not trying to fetch details from GitHub API.');
```

File: RoboFile.php [hunk 3]
```diff
@@ -926,11 +930,38 @@ public function generateReleaseNotes($tag = NULL) {
...
+    }
+
     $log = $this->taskExec("git log --merges --pretty=format:'%s¬¬|¬¬%b' $tag..")->printOutput(FALSE)->run()->getMessage();
```
Comment (amitaibu) [inline RoboFile.php:951]: Worth a comment

File: RoboFile.php [hunk 4]
```diff
@@ -960,12 +1001,118 @@ public function generateReleaseNotes($tag = NULL) {
...
+  protected function githubApiGet($path) {
+    $token = getenv('GITHUB_ACCESS_TOKEN');
+    $username = getenv('GITHUB_USERNAME');
+    if (empty($token)) {
+      throw new Exception('Specify the personal access token in GITHUB_ACCESS_TOKEN environment variable before invoking the release notes generator in order to be able to fetch details of issues and pull requests');
+    }
+    if (empty($username)) {
+      throw new Exception('Specify the GitHub username in GITHUB_USERNAME environment variable before invoking the release notes generator in order to be able to fetch details of issues and pull requests');
+    }
+    // We might not have a sane Drupal instance, let's not rely on Drupal API
```
Comment (amitaibu) [inline RoboFile.php:1103]: :+1:

File: RoboFile.php [hunk 5]
```diff
@@ -960,12 +1001,118 @@ public function generateReleaseNotes($tag = NULL) {
...
+    // to generate release notes.
+    $ch = curl_init('https://api.github.com/' . $path);
+    curl_setopt($ch, CURLOPT_USERAGENT, 'Drupal Starter Release Notes Generator');
+    curl_setopt($ch, CURLOPT_USERPWD, $username . ":" . $token);
+    curl_setopt($ch, CURLOPT_TIMEOUT, 30);
+    curl_setopt($ch, CURLOPT_RETURNTRANSFER, TRUE);
+    $result = curl_exec($ch);
+    curl_close($ch);
+    if (empty($result)) {
```
Comment (amitaibu) [inline RoboFile.php:1112]: `return empty(.. ? ... : ...`

**Review by amitaibu — APPROVED** [review]
:+1:

**Pull request merged**